akka-remote/src/main/scala/akka/remote/artery/tcp/ssl/RotatingKeysSSLEngineProvider.scala

```diff
@@ -0,0 +1,174 @@
+
+/**
+ * Variation on ConfigSSLEngineProvider that will periodically reload the keys and certificates
+ * from disk, to facilitate rolling updates of certificates.
+ *
+ * This class is still ApiMayChange because it can likely be further harmonized with
+ * the standard ConfigSSLEngineProvider. Also the location and default values of the
+ * configuration may change in future versions of Akka.
+ *
+ * This provider does not perform hostname verification, but instead allows checking
+ * that the remote certificate has a subject name that matches the subject name of
+ * the configured certificate.
+ */
+@ApiMayChange
+final class RotatingKeysSSLEngineProvider(val config: Config, protected val log: MarkerLoggingAdapter)
+    extends SSLEngineProvider {
+
+  def this(system: ActorSystem) =
+    this(
+      system.settings.config.getConfig("akka.remote.artery.ssl.rotating-keys-engine"),
+      Logging.withMarker(system, classOf[RotatingKeysSSLEngineProvider].getName))
+
+  // read config
+
+  private val SSLKeyFile: String = config.getString("key-file")
+  private val SSLCertFile: String = config.getString("cert-file")
+  private val SSLCACertFile: String = config.getString("ca-cert-file")
+
+  private val sslEngineConfig = new SSLEngineConfig(config)
+  import sslEngineConfig._
+
+  // build a PRNG (created once, reused on every instance of SSLContext
+  private val rng: SecureRandom = SecureRandomFactory.createSecureRandom(SSLRandomNumberGenerator, log)
+
+  // handle caching
+  @volatile private var cachedContext: Option[CachedContext] = None
+
+  /** INTERNAL API */
+  @InternalApi
+  private[ssl] def getSSLContext() = getContext().context
+  private def getContext(): ConfiguredContext = {
+    cachedContext match {
+      case Some(CachedContext(_, expired)) if expired.isOverdue() =>
+        // Multiple connection requests arriving when the cache is overdue will
+        // create different CachedContext instances and only the last one will
+        // be cached. This is fine.
+        val context = constructContext()
+        cachedContext = Some(CachedContext(context, SSLContextCacheTime.fromNow))
+        context
+      case Some(CachedContext(cached, _)) => cached
+      case None                           =>
+        // Multiple connection requests arriving when the cache is empty will
+        // create different CachedContext instances. This is fine.
+        val context = constructContext()
+        cachedContext = Some(CachedContext(context, SSLContextCacheTime.fromNow))
+        context
+    }
+  }
+
+  // Construct the cached instance
+  private def constructContext(): ConfiguredContext = {
+    val (privateKey, cert, cacert) = readFiles()
+    try {
+      val keyManagers: Array[KeyManager] = PemManagersProvider.buildKeyManagers(privateKey, cert, cacert)
+      val trustManagers: Array[TrustManager] = PemManagersProvider.buildTrustManagers(cacert)
+
+      val sessionVerifier = new PeerSubjectVerifier(cert)
```

Comment:
Perhaps this is not needed in this PR, but it would be nice if the subject verification were configurable. For example, you might turn on/off whether SANs are included or if just the CN is used, and you might also want to specify a name or list of names that are allowed. An example where a list of names from configuration might be valuable is if you had different deployments with different roles, and for whatever reason deployment got their own certificate with their own CN, and you wanted all pods from all of these deployments to form a single cluster.

But, as it is, the peer subject verifier is probably what will be most convenient and useful for most users, so I think this PR is fine as is for now.

Reply:
The thin line between configurable and too-much public API is what kept us busy when porting this code here. The introduction of the `SessionVerifier` trait is leading us into the direction of providing multiple verifiers and being able to configure an array of implementations on settings.

Reply:
created  #29189